ihm/multi_state_scheme.py

```diff
@@ -0,0 +1,225 @@
+        if connectivity.begin_state not in self.states:
+            self.states.append(connectivity.begin_state)
+        if (connectivity.end_state is not None) and \
+                (connectivity.end_state not in self.states):
+            self.states.append(connectivity.end_state)
+
+    def add_relaxation_time(self, relaxation_time):
+        """Add a relaxation time to the scheme. This relaxation time is not
+        assigned to a connectivity.
+
+        :param relaxation_time: The relaxation time to add to the scheme.
+        :type relaxation_time: :class:`RelaxationTime`
+        """
+        self.relaxation_time_list.append(relaxation_time)
+
+    def __eq__(self, other):
+        return ((self.__dict__ == other.__dict__)
+                and (self.connectivity_list == other.connectivity_list)
+                and (self.relaxation_time_list == other.relaxation_time_list))
+
+
+class MultiStateSchemeConnectivity(object):
+    """A connectivity between states. Used to describe the directed
+    edge of graph.
+    If no end_state is given, the state is not connected to another state.
+    This could be the case for states where no connection to other states
+    could be resolved.
+
+    :param begin_state: The start state of the connectivity.
+    :type begin_state: :class:`ihm.model.State`
+    :param end_state: The end state of the connectivity. Can be None in case
+     of states that are not connected to others.
+    :type end_state: :class:`ìhm.model.State`
+    :param details: Details to the connectivity.
+    :param dataset_group: The DatasetGroup that was used to obtain information
+    on the connectivity.
+    :type dataset_group: :class:`ìhm.dataset.DatasetGroup`
+    :param kinetic_rate: A kinetic rate assigned to the connectivity.
+    :type kinetic_rate: :class:`KineticRate`
+    :param relaxation_time: A relaxation time assigned to the connectivity.
+    :type relaxation_time: :class:`RelaxationTime`
+    """
+    def __init__(self, begin_state, end_state=None, details=None,
+                 dataset_group=None, kinetic_rate=None, relaxation_time=None):
+        self.begin_state = begin_state
+        self.end_state = end_state
+        self.details = details
+        self.dataset_group = dataset_group
+        self.kinetic_rate = kinetic_rate
+        self.relaxation_time = relaxation_time
+        # The _assigned_to_scheme variable tracks whether the connectivity
+        # has been assigned to a scheme. This is to ensure that each
+        # connectivity is only assigned to a single scheme.
+        self._assigned_to_scheme = False
+
+    def set_assigned_to_scheme(self):
+        self._assigned_to_scheme = True
+
+    def __eq__(self, other):
+        return self.__dict__ == other.__dict__
+
+
+class KineticRate(object):
+    """A kinetic rate that can be assigned to a connectivity.
+    The kinetic rate could be either a transition_rate_constant or
+    an equilibrium_constant.
```

Comment:
In this case you should have two separate classes that derive from a common base class, perhaps `TransitionKineticRate` and `EquilibriumKineticRate`.

Reply:
In that case what you have already would be fine (with corrected documentation). But consider combining the three equilibrium constant parameters into a separate object. python-ihm convention would also be to use three classes rather than the `equilibrium_constant_determination_method` enumeration, e.g.

```
class EquilibriumConstant(object):
    method = "equilibrium constant is determined from another method not listed"
    def __init__(self, value, unit):
        self.value, self.unit = value, unit

class PopulationEquilibriumConstant(EquilibriumConstant):
    method = "equilibrium constant is determined from population"

class KineticRateEquilibriumConstant(EquilibriumConstant):
    method = "equilibrium constant is determined from kinetic rates, kAB/kBA'"

e = KineticRateEquilibriumConstant(value=0.1, unit="foo")
r = KineticRate(equilibrium_constant=e, ...)
```

That also removes the laborious and error-prone requirement to write out the full value for `equilibrium_constant_determination_method` each time. It is a bit more work in `ihm.reader` to map the enumeration to the correct subclass though.

Reply:
Thank you for your suggestions. I shifted the equilibrium constant part to separate classes accordingly